server/src/main/java/org/opensearch/index/shard/CloseableRetryableRefreshListener.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.shard;
+
+import org.apache.lucene.search.ReferenceManager;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.threadpool.ThreadPool;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * RefreshListener that runs afterRefresh method if and only if there are atleast one permit available. Once the listener
+ * is closed, all the permits are acquired and there are no available permits to afterRefresh. This abstract class provides
+ * necessary abstract methods to schedule retry.
+ */
+public abstract class CloseableRetryableRefreshListener implements ReferenceManager.RefreshListener, Closeable {
+
+    private static final int TOTAL_PERMITS = Integer.MAX_VALUE;
+
+    private final Semaphore semaphore = new Semaphore(TOTAL_PERMITS);
+
+    private final ThreadPool retryThreadPool;
+
+    public CloseableRetryableRefreshListener(ThreadPool retryThreadPool) {
+        this.retryThreadPool = retryThreadPool;
+    }
+
+    @Override
+    public final void afterRefresh(boolean didRefresh) throws IOException {
+        if (semaphore.tryAcquire()) {
+            boolean shouldRetry;
+            try {
+                shouldRetry = performAfterRefresh(didRefresh);
+            } finally {
+                semaphore.release();
+            }
+            scheduleRetry(shouldRetry);
+        }
+    }
+
+    protected String getRetryThreadPoolName() {
+        return null;
+    }
+
+    protected TimeValue getNextRetryInterval() {
+        return null;
+    }
```

Comment:
Make them abstract?

Reply:
Providing default implementation so that the implementor does not need to implement it for cases where retry is not required.